### PR TITLE
Adding support for constructing Dataset from SparseMatrix 

### PR DESCRIPTION
**Pull request by yaxxie**
As title:

- [x] Implement FFI wrapper for creating sparse dataset
- [x] Tests for FFI
- [x] Add support to `fit!`
- [x] Tests for higher level functions
    - [x] Specifically tests to show that the SparseMatrix support is fitting models identically with otherwise identical data
- [ ] [Optional?] Figure out what needs to be done for MLJ
- [ ] [Optional?] Add support for SparseMatrix predictions?

### CODE REVIEW

**Comment by yaxxie** [comment]
@kd-iqvia do you consider this sufficient for an initial take for sparse matrix support?

The tests aren't exhaustive (e.g. I don't demonstrate that you can in fact mix dense/sparse matrices for the train test, and I haven't covered a huge range of scenarios for checking the sparse inputs are correctly received), but it looks like it works. Is there anything you want to see or would need to see before merge?

File: src/wrapper.jl [hunk 1]
```diff
@@ -110,9 +128,40 @@ macro lightgbm(f, params...)
...
+    idx_type, data_type = sparsetypes(data)
+
+    out = Ref{DatasetHandle}()
+    @lightgbm(
+        :LGBM_DatasetCreateFromCSC,
+        data.colptr .- 1 => Ptr{Nothing},
+        idx_type => Cint,
+        Int32.(data.rowval .- 1) => Ptr{Cint},
+        data.nzval => Ptr{Nothing},
+        data_type => Cint,
```
Comment (FatemehTahavori) [inline src/wrapper.jl:154]: data_type – Type of data pointer, can be C_API_DTYPE_FLOAT32 or C_API_DTYPE_FLOAT64 shouldn't be this type?

File: src/wrapper.jl [hunk 2]
```diff
@@ -5,6 +5,8 @@ const C_API_DTYPE_FLOAT32 = 0
 const C_API_DTYPE_FLOAT64 = 1
 const C_API_DTYPE_INT32 = 2
 const C_API_DTYPE_INT64 = 3
+const C_API_MATRIX_TYPE_CSC = 1
+const C_API_MATRIX_TYPE_CSR = 0
```
[the remark this answers is not in the document]
Comment (yaxxie) [reply src/wrapper.jl:9]: Not for this PR but they would get used for the other sparse support API calls, in particular PredictSparse*

File: src/wrapper.jl (hunk 1 above)
Comment (yaxxie) [reply src/wrapper.jl:154]: `data_type` is a variable, which is one of `C_API_DTYPE_FLOAT32` `C_API_DTYPE_FLOAT64` (determined by `sparsedatatype` call, see https://github.com/IQVIA-ML/LightGBM.jl/pull/113/files#diff-e08580f7ddadf1983cf458c25618ce250ee72b50d5924cd1375df3c4b023d963R102) 
`Cint` is the type that needs to be passed to the FFI inteface (under the hood, `ccall`) 

So what we're doing here is passing a C integer, which takes a value encoding F32 or F64, and telling `ccall` that this variable is a C integer

**Pull request merged**